operator/docs/enhancements/zone_aware_replication.md

```diff
@@ -0,0 +1,358 @@
+
+
+//  LokiStackSpec defines the desired state of LokiStack
+type LokiStackSpec struct {
+...
+    // ReplicationFactor defines the policy for log stream replication. (Deprecated: Please use replication.factor instead. This field will be removed in future versions of this CRD)
+    //
+    // +optional
+    // +kubebuilder:validation:Optional
+    // +kubebuilder:validation:Minimum:=1
+    // +operator-sdk:csv:customresourcedefinitions:type=spec,xDescriptors="urn:alm:descriptor:com.tectonic.ui:number",displayName="Replication Factor"
+    ReplicationFactor int32 `json:"replicationFactor,omitempty"`
+	
+    // Replication defines the configuration for Loki data replication
+    //
+    // +optional
+    // +kubebuilder:validation:Optional
+    Replication *ReplicationSpec `json:"replication,omitempty"`
+}
+
+type ReplicationSpec struct {
+    // Factor defines the policy for log stream replication.
+    //
+    // +optional
+    // +kubebuilder:validation:Optional
+    // +kubebuilder:validation:Minimum:=1
+    Factor int32 `json:"factor,omitempty"`
+
+    // Toplogy provides the information on zones in the cluster.
+    //
+    // +optional
+    // +kubebuilder:validation:Optional
+    Topology *TopologySpec
+}
+
+type TopologySpec struct {
+    // Enabled defines a flag to enable/disable the zone awareness flags in teh configuration
+    //
+    // +required
+    // +kubebuilder:validation:Required
+    // +operator-sdk:csv:customresourcedefinitions:type=spec,xDescriptors="urn:alm:descriptor:com.tectonic.ui:booleanSwitch",displayName="Enable"
+    Enabled bool
+
+    // Zone is the key that defines a topology in the Nodes' labels
+    //
+    // +required
+    // +kubebuilder:validation:Required
+    Zones []ZoneSpec
+}
+
+// ZoneSpec defines the spec to support zone aware component deployments.
+type ZoneSpec struct {
+    // MaxSkew describes the maximum degree to which Pods can be unevenly distributed
+    //
+    // +optional
+    // +kubebuilder:validation:optional
+    // +kubebuilder:default:="1"
+    MaxSkew int `json:"maxSkew,omitempty"`
+
+    // Topologykey is the key that defines a topology in the Nodes' labels
+    //
+    // +required
+    // +kubebuilder:validation:Required
+    Topologykey string `json:"topologyKey,omitempty"`
+
+}
+
+```
+
+### Implementation Details/Notes/Constraints
+
+The following manifests represents a full example on how zone-aware data replication be turned on using the `topology.kubernetes.io/zone` node label as a key to spread pods across zones:
+
+```yaml
+apiVersion: loki.grafana.com/v1beta1
+kind: LokiStack
+metadata:
+  name: lokistack-dev
+spec:
+  size: 1x.small
+  storage:
+    secret:
+      name: test
+      type: s3
+  storageClassName: gp2
+  replication:
+    factor: 3
+    topology:
+      enabled: true
+      zones:
+      - key: topology.kubernetes.io/zone
+        maxSkew: 1
+```
+	
+#### General constraints
+
+### Risks and Mitigations
+
+## Design Details
```

Comment:
I think I'd start this section with a summary of the issues we need to solve for the implementation:

- Spreading the pods across failure domains using the Kubernetes scheduler (using the `PodTopologySpreadConstraint`)
- Identifying the failure domain a pod is currently scheduled in (the two solutions discussed in length)
- Relaying the failure-domain information to the Loki configuration (using the `instance_availability_zone` configuration option & dynamically set environment variables)